Sources/GoogleAI/FunctionCalling.swift

```diff
@@ -0,0 +1,235 @@
+  ///     - ``DataType/number``: float, double
+  ///     - ``DataType/string``: enum
+  ///   - description: A brief description of the parameter; may be formatted as Markdown.
+  ///   - nullable: Indicates if the value may be null.
+  ///   - enumValues: Possible values of the element of type ``DataType/string`` with "enum" format.
+  ///     For example, an enum `Direction` may be defined as `["EAST", NORTH", "SOUTH", "WEST"]`.
+  ///   - items: Schema of the elements of type ``DataType/array``.
+  ///   - properties: Properties of type ``DataType/object``.
+  ///   - requiredProperties: Required properties of type ``DataType/object``.
+  public init(type: DataType, format: String? = nil, description: String? = nil,
+              nullable: Bool? = nil,
+              enumValues: [String]? = nil, items: Schema? = nil,
+              properties: [String: Schema]? = nil,
+              requiredProperties: [String]? = nil) {
+    self.type = type
+    self.format = format
+    self.description = description
+    self.nullable = nullable
+    self.enumValues = enumValues
+    self.items = items
+    self.properties = properties
+    self.requiredProperties = requiredProperties
+  }
+}
+
+/// A data type.
+///
+/// Contains the set of OpenAPI [data types](https://spec.openapis.org/oas/v3.0.3#data-types).
+public enum DataType: String, Encodable {
+  /// A `String` type.
+  case string = "STRING"
+
+  /// A floating-point number type.
+  case number = "NUMBER"
+
+  /// An integer type.
+  case integer = "INTEGER"
+
+  /// A boolean type.
+  case boolean = "BOOLEAN"
+
+  /// An array type.
+  case array = "ARRAY"
+
+  /// An object type.
+  case object = "OBJECT"
+}
+
+/// Structured representation of a function declaration.
+///
+/// This `FunctionDeclaration` is a representation of a block of code that can be used as a ``Tool``
+/// by the model and executed by the client.
+public struct FunctionDeclaration {
+  /// The name of the function.
+  let name: String
+
+  /// A brief description of the function.
+  let description: String
+
+  /// Describes the parameters to this function; must be of type ``DataType/object``.
+  let parameters: Schema?
+
+  /// Constructs a new `FunctionDeclaration`.
+  ///
+  /// - Parameters:
+  ///   - name: The name of the function; must be a-z, A-Z, 0-9, or contain underscores and dashes,
+  ///   with a maximum length of 63.
+  ///   - description: A brief description of the function.
+  ///   - parameters: Describes the parameters to this function; the keys are parameter names and
+  ///   the values are ``Schema`` objects describing them.
+  ///   - requiredParameters: A list of required parameters by name.
+  public init(name: String, description: String, parameters: [String: Schema]?,
+              requiredParameters: [String]?) {
+    self.name = name
+    self.description = description
+    self.parameters = Schema(
+      type: .object,
+      properties: parameters,
```

Comment:
Should it be consistently named parameters?

Reply:
Potentially in a follow-up. I'm leaning towards keeping it as `properties` to align with [Android](https://github.com/google/generative-ai-android/blob/a2b71dfbc3c05dbbd80cf9148c1cf0a46a40f911/common/src/main/kotlin/com/google/ai/client/generativeai/common/client/Types.kt#L38-L47) and the [REST API](https://ai.google.dev/api/rest/v1beta/Tool?_gl=1*8ma6q8*_up*MQ..*_ga*MTI4MDEwMjkwNC4xNzExNDg2MTc1*_ga_P1DBVKWT6V*MTcxMTQ4NjE3NS4xLjAuMTcxMTQ4NjE3NS4wLjAuMA..#schema) for now, if that seems reasonable.

Reply:
SGTM